README.md
# mocaccino

[![Node.js CI](https://github.com/mantoni/mocaccino.js/actions/workflows/test.yml/badge.svg)](https://github.com/mantoni/mocaccino.js/actions/workflows/test.yml)
[![SemVer]](https://semver.org)
[![License]](https://github.com/mantoni/mocaccino.js/blob/master/LICENSE)

[Mocha][] test runner as a [Browserify][] plugin.

## Install

```
npm install mocaccino --save-dev
```

## Real browser testing

This module is developed as part of [Mochify][] which allows you to run tests
with a headless browser, on a Selenium grid, in the cloud with SauceLabs or
generates a standalone HTML page to run the tests. The underlying modules can
also be used as Browserify plugins.

- [min-webdriver][] for Selenium and SauceLabs support
- [Consolify][] to generate a standalone HTML page

### Code coverage with node

Use the [Coverify][] transform and node:

```
$ browserify --bare -p [ mocaccino --node ] -t coverify test.js | node | coverify
```

## Usage

Mocaccino is a browserify plugin:

```
browserify -p [ mocaccino OPTIONS ]

where OPTIONS are:

  --reporter, -R  Mocha reporter to use, defaults to "tap"
  --grep          Mocha grep option
  --fgrep         Mocha fgrep option
  --invert        Mocha invert option
  --timeout, -t   Mocha timeout in milliseconds to use, defaults to 2000
  --ui, -U        Mocha user interface to use, defaults to "bdd"
  --yields, -y    Yield every N milliseconds, defaults to 250
  --node          If result is used in node instead of a browser
  --windowWidth   Overrides the window width, defaults to the current shells
                  window width or fall back to 80
  --no-colors     Disable colors (overrides color support detection)
  --colors        Enable colors (overrides color support detection)
  --mochaPath     Path to custom Mocha module
```

The `yields` option causes a tiny delay every N milliseconds to allow pending
I/O to happen. It's ignored if `--node` is given.

## Compatibility

- Node 12 or later
- Browserify 5.9 or later (since version 1.0.0)
- Browserify 4.x (before 1.0.0)

## License

MIT

[Build Status]: http://img.shields.io/travis/mantoni/mocaccino.js.svg
[SemVer]: http://img.shields.io/:semver-%E2%9C%93-brightgreen.svg
[License]: http://img.shields.io/npm/l/mocaccino.svg
[Mocha]: http://mochajs.org/
[Browserify]: http://browserify.org
[Mochify]: https://github.com/mantoni/mochify.js
[min-webdriver]: https://github.com/mantoni/min-webdriver
[Consolify]: https://github.com/mantoni/consolify
[Coverify]: https://github.com/substack/coverify


CHANGES.md
# Changes

## 6.1.0

- 🛡 [`fc6f1a5`](https://github.com/mantoni/mocaccino.js/commit/fc6f1a5595e54939c15ce53e5279d059bdeddc91)
  Update mocha to 9.2.0 (#40) (Frederik Ring)

_Released by Frederik Ring on 2022-01-24._

## 6.0.0

- 💥 [`4837604`](https://github.com/mantoni/mocaccino.js/commit/4837604058da7596514f12668c99fa784e8f3535)
  Update mocha to 9.1.3 (#39) (Albert Wang)
- 🛡 [`09f1a80`](https://github.com/mantoni/mocaccino.js/commit/09f1a80bede9e55279160f67edb878499185fc22)
  Update npm dependencies (#38) (Albert Wang)
    * Update eslint dependency to 8.5.0
    * Update through2 dependency to 4.0.2
    * Update supports-color dependency to v8.1.1
    * Update puppeteer dependency to v13
- 🛡 [`37403dc`](https://github.com/mantoni/mocaccino.js/commit/37403dce33af9fbf61645470211da1262f42a3f2)
  Set up GitHub Actions (#37) (Albert Wang)
    * Set up Github Actions
    * Switch CI badge from Travis to Github Actions
    * Recreate package-lock.json
- 🛡 [`7e0d52d`](https://github.com/mantoni/mocaccino.js/commit/7e0d52db43e5d43ccabdabd92067f08640781514)
  Fix test (#36) (Albert Wang)

_Released by Frederik Ring on 2021-12-28._

## 5.0.0

- 💥 [`2a6c8c9`](https://github.com/mantoni/mocaccino.js/commit/2a6c8c922765bb3ca01abd541ac40f24049bfcf2)
  Update mocha to v8, adjust assertions (Frederik Ring)
- 🍏 [`a90d13a`](https://github.com/mantoni/mocaccino.js/commit/a90d13a98b9db38be7bb625459a8ce7c88d1a764)
  Port browser tests to use puppeteer (Frederik Ring)
- 🐛 [`4382056`](https://github.com/mantoni/mocaccino.js/commit/4382056553ec27774483e8cac3d884e8dae31f9a)
  Ensure brout is required before mocha (Frederik Ring)
- 🐛 [`3fa740a`](https://github.com/mantoni/mocaccino.js/commit/3fa740a258b02bdfb2e0783fec4520a059a445da)
  Pass in custom reporters in setup to work around dynamic require limitation (Frederik Ring)
- 🛡 [`05ee03e`](https://github.com/mantoni/mocaccino.js/commit/05ee03ecd40ac9117c2aa490da8938610ebd542c)
  npm audit
- 🛡 [`7f1ae36`](https://github.com/mantoni/mocaccino.js/commit/7f1ae368f9732a270b95b20a788d3c82093980a1)
  Bump lodash from 4.17.19 to 4.17.21 (dependabot[bot])
- 🛡 [`637cb13`](https://github.com/mantoni/mocaccino.js/commit/637cb13990bbdabee46b4ee9d3ae2b8ae23d6235)
  Bump elliptic from 6.4.1 to 6.5.4 (dependabot[bot])
- 🛡 [`9a81451`](https://github.com/mantoni/mocaccino.js/commit/9a81451b806bcf3dc445b2927c8edd73f7ddc029)
  Bump js-yaml from 3.12.0 to 3.14.0 (dependabot[bot])
- 🛡 [`2fce39a`](https://github.com/mantoni/mocaccino.js/commit/2fce39a82b3690ae1f20e5814c8925b0097a16f4)
  Bump lodash from 4.17.11 to 4.17.19 (dependabot[bot])
- 📚 [`c29813f`](https://github.com/mantoni/mocaccino.js/commit/c29813fb727b86832324b3592c855e9039b89ec8)
  Remove phantomic references from README and update engine requirements (Frederik Ring)
- ✨ [`953e4ee`](https://github.com/mantoni/mocaccino.js/commit/953e4ee032af83c5b4e04b215cea557c8cc27316)
  Upgrade Studio Changes

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2021-05-23._

## 4.1.2

- [`f46487a`](https://github.com/mantoni/mocaccino.js/commit/f46487a16d1da04bcad60d111ea98b900e2b62e4)
  Do not assume `process.nextTick` exists

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2019-10-09._

## 4.1.1

- [`fc0eb96`](https://github.com/mantoni/mocaccino.js/commit/fc0eb96deeaeb57d0f7c515177a1bfa418fcd837)
  Fix problems with absolute custom mocha path (#25) (Christopher Hiller)

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2019-03-12._

## 4.1.0

Adds the `mochaPath` option to specify a custom path to resolve mocha.

- [`aea12b7`](https://github.com/mantoni/mocaccino.js/commit/aea12b70e0540a44bc624999997c7f13282fdd1e)
  Add support for custom path to mocha module (Christopher Hiller) (#24)
- [`40e28dc`](https://github.com/mantoni/mocaccino.js/commit/40e28dce56e89faed33398294cf3d00afce21d88)
  Add contributor

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2019-02-28._

## 4.0.0

- [`1b3d32a`](https://github.com/mantoni/mocaccino.js/commit/1b3d32ade806842fb10d94985612e58be76e8d0d)
  Show build status on master branch in README
- [`575abdc`](https://github.com/mantoni/mocaccino.js/commit/575abdc58975954d2f59fd72a156351c3e63c375)
  Run builds in Node 6, 8 and 10
- [`8cfa0ba`](https://github.com/mantoni/mocaccino.js/commit/8cfa0ba7d60ca906a67da3fb0bf3c12edf197afe)
  Replace jslint with eslint
- [`da0d287`](https://github.com/mantoni/mocaccino.js/commit/da0d287602278aa59b8b6034939ceb819a0d99ee)
  Update all transient dependencies
- [`ca8e25c`](https://github.com/mantoni/mocaccino.js/commit/ca8e25c3fa0b222117b765c7d45d8242e4badc71)
  Update resolve and through2
- [`3f3317d`](https://github.com/mantoni/mocaccino.js/commit/3f3317d5ea20c58e19ae032c954aa8c8b98d93f2)
  Upgrade Mocha to v5
- [`3bea39e`](https://github.com/mantoni/mocaccino.js/commit/3bea39eb5f2b278dcaa76f4ac6d04cf46fc20e34)
  Upgrade supports-color to v5
- [`db00c21`](https://github.com/mantoni/mocaccino.js/commit/db00c21b5a494c9e5ec41806469ccc1a6fa5f426)
  Upgrade Browserify to v16
- [`84dd99a`](https://github.com/mantoni/mocaccino.js/commit/84dd99ad718954f614f0327bb772ed377b7b9891)
  Update @studio/changes for `--commits` and `--footer` support

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2018-11-19._

## 3.0.0

- Update to Mocha 4 (Carl-Erik Kopseng)

## 2.0.0

- Upgrade Mocha to `^3.2` (Morgan Roderick)
- Order dependencies by using `npm` command (Morgan Roderick)

    This will make future changesets made with `npm install` read much better,
    as `npm` commands always sorts dependencies lexicographically.

- Run tests in newer node versions, and remove 0.10, 0.12 (#21) (Morgan
  Roderick)
- Use `@studio/changes` instead of make

## 1.9.0

- Add support for `reporterOptions` in API (Tomer Lahav)
- Update Mocha link (Morton Fox)

## 1.8.2

Revert "Nail Mocha to `~2.3.4` to workaround issues with `2.4.x`" since the
underlying issue was reverted in Mocha.

## 1.8.1


test/fixture/exit.js
/*
 * mocaccino.js
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
/*globals window, describe, it*/
'use strict';

window.process = {
  exit: function () {
    console.log('Mocaccino FTW!');
  }
};

describe('fixture', function () {

  it('passes', function () {
    return;
  });

});


test/fixture/fail.js
/*
 * mocaccino.js
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
/*globals describe, it*/
'use strict';

describe('fixture', function () {

  it('fails', function () {
    throw new Error('Ouch!');
  });

});



test/fixture/pass.js
/*
 * mocaccino.js
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
/*globals describe, it*/
'use strict';

describe('fixture', function () {

  it('passes', function () {
    return;
  });

});


lib/setup-node.js
/*
 * mocaccino.js
 *
 * Copyright (c) 2014-2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var Mocha = require('mocha');
Mocha.reporters.Base.window.width = JSON.parse('{{WINDOW_WIDTH}}');
Mocha.reporters.Base.symbols.dot = '.';
var grep = '{{GREP}}';
var fgrep = '{{FGREP}}';
var _mocha = new Mocha({
  grep: grep.length ? grep : undefined,
  fgrep: fgrep.length ? fgrep : undefined
});
_mocha.ui('{{UI}}');
if ('{{INVERT}}' === true) { // eslint-disable-line no-constant-condition
  _mocha.invert();
}
_mocha.reporter('{{REPORTER}}', '{{REPORTER_OPTIONS}}');
_mocha.timeout('{{TIMEOUT}}');
_mocha.color('{{USE_COLORS}}');
_mocha.suite.emit('pre-require', global, '', _mocha);

setTimeout(function () {
  _mocha.run(function (errs) {
    process.exit(errs ? 1 : 0);
  });
}, 1);


lib/setup-browser.js
/*global Mocha, window*/
/*
 * mocaccino.js
 *
 * Copyright (c) 2014-2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

Mocha.reporters.Base.window.width = JSON.parse('{{WINDOW_WIDTH}}');
Mocha.reporters.Base.symbols.dot = '.';
var grep = '{{GREP}}';
var fgrep = '{{FGREP}}';
var mocha = new Mocha({
  grep: grep.length ? grep : undefined,
  fgrep: fgrep.length ? fgrep : undefined
});
if ('{{INVERT}}' === true) { // eslint-disable-line no-constant-condition
  mocha.invert();
}

if (Mocha.reporters['{{REPORTER}}']) {
  mocha.reporter('{{REPORTER}}', '{{REPORTER_OPTIONS}}');
} else {
  mocha.reporter(require('{{REPORTER}}'), '{{REPORTER_OPTIONS}}');
}

mocha.ui('{{UI}}');
mocha.timeout('{{TIMEOUT}}');
mocha.color('{{USE_COLORS}}');
mocha.suite.emit('pre-require', window, '', mocha);
var t = new Date().getTime();
var y = Number('{{YIELDS}}');
mocha.suite.afterEach(function (done) {
  var now = new Date().getTime();
  if (now - t > y) {
    t = now;
    if (process.nextTick) {
      process.nextTick(done);
    } else {
      setTimeout(done, 0);
    }
  } else {
    done();
  }
});

setTimeout(function () {
  Mocha.process.stdout = process.stdout;
  mocha.run(function (errs) {
    process.exit(errs ? 1 : 0);
  });
}, 1);


lib/mocaccino.js
/*
 * mocaccino.js
 *
 * Copyright (c) 2014-2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var path    = require('path');
var fs      = require('fs');
var through = require('through2');
var resolve = require('resolve');
var listen  = require('listen');
var supportsColor = require('supports-color');

module.exports = function (b, opts) {
  if (!opts) {
    opts = {};
  }
  var reporter = opts.reporter || opts.R || 'tap';
  var reporterOptions = opts.reporterOptions || {};
  var grep = opts.grep instanceof RegExp
    ? opts.grep.source
    : (opts.grep || '');
  var fgrep = opts.fgrep || '';
  var invert = opts.invert;
  var yields = opts.yields || opts.y || 250;
  var ui = opts.ui || opts.U || 'bdd';
  var timeout = opts.timeout || opts.t || 2000;
  var colors = typeof opts.colors === 'boolean'
    ? opts.colors
    : supportsColor.stdout;
  var setupFile;
  var listener = listen();
  if (opts.node) {
    b.exclude('mocha');
    setupFile = 'setup-node.js';
  } else {
    setupFile = 'setup-browser.js';
    var broutFile = resolve.sync('brout', {
      baseDir: __dirname,
      packageFilter: function (pkg) {
        return { main : pkg.browser };
      }
    });
    var broutPath = path.relative(process.cwd(), broutFile);
    b.require('./' + broutPath.replace(/\\/g, '/'), {
      expose : 'brout'
    });

    var mochaReporters = require(opts.mochaPath
      ? path.resolve(process.cwd(), opts.mochaPath, 'lib', 'mocha')
      : 'mocha').reporters;
    if (!mochaReporters[reporter]) {
      var reporterFile = resolve.sync(reporter, {
        baseDir: __dirname
      });
      b.add(reporterFile);
      b.require(reporterFile, {
        expose: reporter
      });
    }

    var mochaPath = opts.mochaPath
      ? resolve.sync('mocha.js', { paths: [opts.mochaPath] })
      : resolve.sync('mocha/mocha');
    fs.readFile(mochaPath, 'utf8', listener('mocha'));
  }
  setupFile = path.join(__dirname, setupFile);

  var onSetupFile = listener('setup');
  fs.readFile(setupFile, 'utf8', function (err, content) {
    if (err) {
      throw err;
    }
    var windowWidth = opts.windowWidth
      || (process.stdout.getWindowSize && process.stdout.getWindowSize()[0])
      || 80;
    var setupContent = content
      .replace('{{WINDOW_WIDTH}}', JSON.stringify(windowWidth))
      .replace(/{{REPORTER}}/g, reporter)
      .replace(/'{{REPORTER_OPTIONS}}'/g, JSON.stringify(reporterOptions))
      .replace('{{YIELDS}}', yields)
      .replace('{{UI}}', ui)
      .replace('\'{{TIMEOUT}}\'', timeout)
      .replace('\'{{USE_COLORS}}\'', colors ? 'true' : 'false')
      .replace('{{GREP}}', grep)
      .replace('{{FGREP}}', fgrep)
      .replace('\'{{INVERT}}\'', invert ? 'true' : 'false');
    onSetupFile(null, setupContent);
  });

  if (!opts.node) {
    b.transform(function () {
      var s = '';
      return through(function (chunk, enc, next) {
        s += chunk;
        var p = s.lastIndexOf('\n');
        if (p !== -1) {
          this.push(
            s.substring(0, p + 1).replace(/require\(['"]mocha["']\)/g, 'Mocha')
          );
          s = s.substring(p + 1);
        }
        next();
      }, function (next) {
        if (s) {
          this.push(s);
        }
        next();
      });
    });
  }

  var setupSource;
  var reset = false;

  function createSetupEntry() {
    return {
      id     : 'mocaccino-setup',
      file   : setupFile,
      entry  : true,
      order  : 0,
      source : setupSource,
      deps   : {}
    };
  }

  function apply() {
    b.pipeline.get('deps').push(through.obj(function (row, enc, next) {
      if (row.entry) {
        row.deps['mocaccino-setup'] = setupFile;
        row.order++;
      }
      this.push(row);
      if (listener) {
        var self = this;
        listener.then(function (err, res) {
          if (!err) {
            setupSource = (opts.node ? '' : 'require(\'brout\');\n')
              + (res.mocha || '')
              + res.setup;
            self.push(createSetupEntry());
          }
          next();
        });
        listener = null;
      } else if (reset) {
        this.push(createSetupEntry());
        reset = false;
        next();
      } else {
        next();
      }
    }));
  }

  apply();
  b.on('reset', function () {
    reset = true;
    apply();
  });

};
